Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,349 @@
+                var al = new ActiveListener<T>(gl);
+                // add each of them to the listeners list
+                // if it is the first one, a new List must be created and the object be added
+                // otherwise add it to the list
+                listeners.AddOrUpdate(uniqueId,
+                    // add new list
+                    i =>
+                    {
+                        var l = new List<ActiveListener<T>>();
+                        l.Add(al);
+                        return l;
+                    },
+                    // update list
+                    (i, l) =>
+                    {
+                        l.Add(al);
+                        return l;
+                    }
+                );
+            }
+
+            // execute all listeners async
+            await Task.WhenAll(listeners[uniqueId].Select(t => t.Listener(thing)));
+            // save changes
+            await thing.SaveChanges();
+        }
+
+        public sealed override Task OnUserDisconnected(IUser user)
+        {
+            return OnThingRemoved(UserListeners, user.SessionId);
+        }
+
+        public override Task OnChannelDeleted(IChannel channel)
+        {
+            return OnThingRemoved(ChannelListeners, channel.ChannelId);
+        }
+
+        private Task OnThingRemoved<T>(ConcurrentDictionary<int, IList<ActiveListener<T>>> listeners, int uniqueId) where T : IReadModifyWriteObject
+        {
+            IList<ActiveListener<T>> dontCare;
+            listeners.TryRemove(uniqueId, out dontCare);
+            return Task.FromResult<object>(null);
+        }
+
+        public sealed override Task OnUserModified(IUser user)
+        {
+            return OnThingModified(UserListeners, user, user.SessionId);
+        }
+
+        public sealed override Task OnChannelModified(IChannel channel)
+        {
+            return OnThingModified(ChannelListeners, channel, channel.ChannelId);
+        }
+
+        private async Task OnThingModified<T>(ConcurrentDictionary<int, IList<ActiveListener<T>>> listeners, T thing, int uniqueId) where T : IReadModifyWriteObject
+        {
+            // every time a user is modified, all listening functions with changed objects need to be called
+            // get all listening functions of the user
+            await Task.WhenAll(listeners[uniqueId]
+                // check if function has recently been added (Item3 == null)
+                // or if the old saved object differs from the newly calculated (and directly saved) one
+                .Where(t => {
+                    var prev = t.Obj;
+                    t.Obj = t.Mapper(thing);
+                    return null == prev || !prev.Equals(t.Obj);
+                })
+                // get all listener functions
+                .Select(t => t.Listener(thing))
+            );
+
+            // save the user after all Tasks have been finished
+            await thing.SaveChanges();
```

Comment:
This is kinda wasteful, as listeners needn't change stuff.

Actually, they really *shouldn't* (infinite loop if someone has a listener on a property and modifies that property within the listener).